### ISSUE TITLE
DMA: Out of SW-IOMMU space for 9100 bytes at device 

### ISSUE THREAD

**Issue by taolinke** [lwfinger/rtlwifi_new#30]
hey,i got so many "DMA: Out of SW-IOMMU space for 9100 bytes at device" in /var/log/messages,then my laptop not answering.
my wireless card is rtl8192ee.
03:00.0 Network controller: Realtek Semiconductor Co., Ltd. RTL8192EE PCIe Wireless Network Adapter

**Comment by lambertdw** [comment]
Hi,

I'm using the  product: RTL8192EE PCIe Wireless Network Adapter with the 
   lwfinger/rtlwifi_new code.  It's wonderful because I no longer need a 
wired connection.
However, the connection drops frequently (for instance, 3 attempts 
before successfully transferring a few photos via airdroid).  Which code 
should I use?  How can I help debug the driver?
I am fluent in C, and unix as an end user.  I haven't worked with device 
drivers since reading the assembly code published for the original IBM 
PC.  Installing the 14.10 would not trouble me if you me advised to do so.
<https://github.com/lwfinger/rtlwifi_new>

This system is lenovo W540 Ubuntu linux distribution long term 14.04
$ uname -a
Linux dwl 3.13.0-45-generic #74-Ubuntu SMP Tue Jan 13 19:36:28 UTC 2015 
x86_64 x86_64 x86_64 GNU/Linux
$ cat /proc/version_signature
Ubuntu 3.13.0-45.74-generic 3.13.11-ckt13
$

Thank you,
Dave Lambert

**Comment by lwfinger** [comment]
I hope you did a 'git clone' on the rtlwifi_new repo. If so, then 'git checkout troy' and try that branch.

**Comment by lambertdw** [comment]
I  forked these repositories.  Anyway, I believe I've installed the troy 
branch, and it seems improved, this airdroid test was far more (though 
not completely) successful.
ftp from old computer to new computer via the LAN works poorly---manages 
to disconnect old trusty as well as new computer.
I set up an unencrypted hidden network, the transfer rate is 1.4 
MiB/second, looking solid.

hmm, complete system (W540) crash when I reconnected to the dlink router.

(old system is standard Lenovo T500 ubuntu 14.10 with failing hinges, 
keyboard and more.  Wireless router is dlink 655.)
Hope this information helps.

On 02/04/2015 12:21 AM, lwfinger wrote:
>
> I hope you did a 'git clone' on the rtlwifi_new repo. If so, then 'git 
> checkout troy' and try that branch.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/lwfinger/rtlwifi_new/issues/30#issuecomment-72792565>.
>

-- 
We plan to eliminate road runner service, please change my address destination to gmail.com .
Thank you, your lover/family/friend/acquaintance Dave

**Comment by kevinmullican** [comment]
I am also seeing this issue working off the 'rtlwifi_new' repo, 'master' branch. I am primarily using hostapd on this machine. I'm more than happy to help debug. The system is as follows:

Gigabyte Brix GB-BXi7-4770R

~/src$ uname -a
Linux minimezz 3.16.0-31-generic #43~14.04.1-Ubuntu SMP Tue Mar 10 20:13:38 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

~/src$ more /etc/issue
Ubuntu 14.04.2 LTS \n \l

~/src$ lspci | grep Network
02:00.0 Network controller: Realtek Semiconductor Co., Ltd. RTL8821AE 802.11ac PCIe Wireless Network Adapter

~/src$ lsmod | grep rtl
rtl8821ae             318773  0 
btcoexist             183378  1 rtl8821ae
rtl_pci                39740  1 rtl8821ae
rtlwifi               121320  3 btcoexist,rtl_pci,rtl8821ae
mac80211              652718  3 rtl_pci,rtlwifi,rtl8821ae
cfg80211              494330  2 mac80211,rtlwifi

The relevant syslog entries look like:

Mar 13 19:28:36 minimezz kernel: [ 3627.469443] rtlwifi:rtl_cam_del_entry():<0-0> cam_bitmap: 0x4
Mar 13 19:28:36 minimezz kernel: [ 3627.469484] rtlwifi:rtl_cam_get_free_entry():<0-0> key_index=1,cam_bitmap: 0x6 entry_idx=1
Mar 13 19:36:36 minimezz kernel: [ 4108.375773] rtl8821ae 0000:02:00.0: swiotlb buffer is full (sz: 9100 bytes)
Mar 13 19:36:36 minimezz kernel: [ 4108.375776] DMA: Out of SW-IOMMU space for 9100 bytes at device 0000:02:00.0
Mar 13 19:36:36 minimezz kernel: [ 4108.395457] rtl8821ae 0000:02:00.0: swiotlb buffer is full (sz: 9100 bytes)
Mar 13 19:36:36 minimezz kernel: [ 4108.395459] DMA: Out of SW-IOMMU space for 9100 bytes at device 0000:02:00.0
Mar 13 19:36:37 minimezz kernel: [ 4109.004580] rtl8821ae 0000:02:00.0: swiotlb buffer is full (sz: 9100 bytes)
Mar 13 19:36:37 minimezz kernel: [ 4109.004582] DMA: Out of SW-IOMMU space for 9100 bytes at device 0000:02:00.0
Mar 13 19:36:43 minimezz kernel: [ 4115.198090] rtl8821ae 0000:02:00.0: swiotlb buffer is full (sz: 9100 bytes)
Mar 13 19:36:43 minimezz kernel: [ 4115.198092] DMA: Out of SW-IOMMU space for 9100 bytes at device 0000:02:00.0
Mar 13 19:36:43 minimezz kernel: [ 4115.260206] rtl8821ae 0000:02:00.0: swiotlb buffer is full (sz: 9100 bytes)
Mar 13 19:36:43 minimezz kernel: [ 4115.260209] DMA: Out of SW-IOMMU space for 9100 bytes at device 0000:02:00.0
Mar 13 19:38:23 minimezz kernel: [ 4214.892517] rtl8821ae 0000:02:00.0: swiotlb buffer is full (sz: 9100 bytes)
Mar 13 19:38:23 minimezz kernel: [ 4214.892524] DMA: Out of SW-IOMMU space for 9100 bytes at device 0000:02:00.0
Mar 13 19:38:23 minimezz kernel: [ 4214.927682] rtl8821ae 0000:02:00.0: swiotlb buffer is full (sz: 9100 bytes)
Mar 13 19:38:23 minimezz kernel: [ 4214.927685] DMA: Out of SW-IOMMU space for 9100 bytes at device 0000:02:00.0
Mar 13 19:38:23 minimezz kernel: [ 4215.306340] rtl8821ae 0000:02:00.0: swiotlb buffer is full (sz: 9100 bytes)
Mar 13 19:38:23 minimezz kernel: [ 4215.306342] DMA: Out of SW-IOMMU space for 9100 bytes at device 0000:02:00.0
Mar 13 19:38:25 minimezz kernel: [ 4217.311950] rtl8821ae 0000:02:00.0: swiotlb buffer is full (sz: 9100 bytes)
Mar 13 19:38:25 minimezz kernel: [ 4217.311952] DMA: Out of SW-IOMMU space for 9100 bytes at device 0000:02:00.0
Mar 13 19:38:26 minimezz kernel: [ 4217.690910] rtl8821ae 0000:02:00.0: swiotlb buffer is full (sz: 9100 bytes)
Mar 13 19:38:26 minimezz kernel: [ 4217.690913] DMA: Out of SW-IOMMU space for 9100 bytes at device 0000:02:00.0
Mar 13 19:38:27 minimezz kernel: [ 4219.042106] rtl8821ae 0000:02:00.0: swiotlb buffer is full (sz: 9100 bytes)
Mar 13 19:38:27 minimezz kernel: [ 4219.042109] DMA: Out of SW-IOMMU space for 9100 bytes at device 0000:02:00.0
Mar 13 19:38:27 minimezz kernel: [ 4219.317534] rtl8821ae 0000:02:00.0: swiotlb buffer is full (sz: 9100 bytes)
Mar 13 19:38:27 minimezz kernel: [ 4219.317537] DMA: Out of SW-IOMMU space for 9100 bytes at device 0000:02:00.0
Mar 13 19:38:31 minimezz kernel: [ 4222.975079] rtl8821ae 0000:02:00.0: swiotlb buffer is full (sz: 9100 bytes)
Mar 13 19:38:31 minimezz kernel: [ 4222.975082] DMA: Out of SW-IOMMU space for 9100 bytes at device 0000:02:00.0
Mar 13 19:38:31 minimezz kernel: [ 4222.990949] rtl8821ae 0000:02:00.0: swiotlb buffer is full (sz: 9100 bytes)
Mar 13 19:38:31 minimezz kernel: [ 4222.990951] DMA: Out of SW-IOMMU space for 9100 bytes at device 0000:02:00.0
Mar 13 19:38:31 minimezz kernel: [ 4223.328800] rtl8821ae 0000:02:00.0: swiotlb buffer is full (sz: 9100 bytes)
Mar 13 19:38:31 minimezz kernel: [ 4223.328802] DMA: Out of SW-IOMMU space for 9100 bytes at device 0000:02:00.0
Mar 13 19:38:33 minimezz kernel: [ 4225.334401] rtl8821ae 0000:02:00.0: swiotlb buffer is full (sz: 9100 bytes)
Mar 13 19:38:33 minimezz kernel: [ 4225.334404] DMA: Out of SW-IOMMU space for 9100 bytes at device 0000:02:00.0
Mar 13 19:38:33 minimezz kernel: [ 4225.468351] rtl8821ae 0000:02:00.0: swiotlb buffer is full (sz: 9100 bytes)
Mar 13 19:38:33 minimezz kernel: [ 4225.468353] DMA: Out of SW-IOMMU space for 9100 bytes at device 0000:02:00.0
Mar 13 19:38:36 minimezz kernel: [ 4227.956618] rtlwifi:rtl_cam_del_entry():<0-0> cam_bitmap: 0x2
Mar 13 19:38:36 minimezz kernel: [ 4227.956653] rtlwifi:rtl_cam_get_free_entry():<0-0> key_index=2,cam_bitmap: 0x6 entry_idx=2
Mar 13 19:38:59 minimezz kernel: [ 4251.500166] rtl8821ae 0000:02:00.0: swiotlb buffer is full (sz: 9100 bytes)
Mar 13 19:38:59 minimezz kernel: [ 4251.500168] DMA: Out of SW-IOMMU space for 9100 bytes at device 0000:02:00.0
Mar 13 19:39:01 minimezz kernel: [ 4253.417076] rtl8821ae 0000:02:00.0: swiotlb buffer is full (sz: 9100 bytes)

**Comment by kevinmullican** [comment]
Just a quick update. The module built from the 'troy' branch fails to load. The machine hangs indefinitely on modprobe.

**Comment by lwfinger** [comment]
The troy branch works fine for me. I have no idea what is causing your problem.

**Comment by kevinmullican** [comment]
alright. I've gone back to the master branch. the machine crashes hard
after several hours of running. the following errors are in the syslog just
before the crash:

Mar 16 18:11:18 minimezz kernel: [ 4828.548678]
rtlwifi:rtl_cam_del_entry():<0-0> cam_bitmap: 0x4
Mar 16 18:11:18 minimezz kernel: [ 4828.548735]
rtlwifi:rtl_cam_get_free_entry():<0-0> key_index=1,cam_bitmap: 0x6
entry_idx=1
Mar 16 18:11:41 minimezz kernel: [ 4851.947037] rtl8821ae 0000:02:00.0:
swiotlb buffer is full (sz: 9100 bytes)
Mar 16 18:11:41 minimezz kernel: [ 4851.947039] DMA: Out of SW-IOMMU space
for 9100 bytes at device 0000:02:00.0
Mar 16 18:11:42 minimezz kernel: [ 4852.041375] rtl8821ae 0000:02:00.0:
swiotlb buffer is full (sz: 9100 bytes)
Mar 16 18:11:42 minimezz kernel: [ 4852.041377] DMA: Out of SW-IOMMU space
for 9100 bytes at device 0000:02:00.0
Mar 16 18:11:42 minimezz kernel: [ 4852.043987] rtl8821ae 0000:02:00.0:
swiotlb buffer is full (sz: 9100 bytes)
Mar 16 18:11:42 minimezz kernel: [ 4852.043989] DMA: Out of SW-IOMMU space
for 9100 bytes at device 0000:02:00.0
Mar 16 18:11:42 minimezz kernel: [ 4852.134972] rtl8821ae 0000:02:00.0:
swiotlb buffer is full (sz: 9100 bytes)
Mar 16 18:11:42 minimezz kernel: [ 4852.134974] DMA: Out of SW-IOMMU space
for 9100 bytes at device 0000:02:00.0
Mar 16 18:11:42 minimezz kernel: [ 4852.147478] rtl8821ae 0000:02:00.0:
swiotlb buffer is full (sz: 9100 bytes)
Mar 16 18:11:42 minimezz kernel: [ 4852.147480] DMA: Out of SW-IOMMU space
for 9100 bytes at device 0000:02:00.0
Mar 16 18:11:42 minimezz kernel: [ 4852.148817] rtl8821ae 0000:02:00.0:
swiotlb buffer is full (sz: 9100 bytes)
Mar 16 18:11:42 minimezz kernel: [ 4852.148819] DMA: Out of SW-IOMMU space
for 9100 bytes at device 0000:02:00.0
Mar 16 18:11:42 minimezz kernel: [ 4852.151160] rtl8821ae 0000:02:00.0:
swiotlb buffer is full (sz: 9100 bytes)
Mar 16 18:11:42 minimezz kernel: [ 4852.151162] DMA: Out of SW-IOMMU space
for 9100 bytes at device 0000:02:00.0
Mar 16 18:11:42 minimezz kernel: [ 4852.153765] rtl8821ae 0000:02:00.0:
swiotlb buffer is full (sz: 9100 bytes)
Mar 16 18:11:42 minimezz kernel: [ 4852.153767] DMA: Out of SW-IOMMU space
for 9100 bytes at device 0000:02:00.0
Mar 16 18:11:42 minimezz kernel: [ 4852.248060] rtl8821ae 0000:02:00.0:
swiotlb buffer is full (sz: 9100 bytes)
Mar 16 18:11:42 minimezz kernel: [ 4852.248062] DMA: Out of SW-IOMMU space
for 9100 bytes at device 0000:02:00.0
Mar 16 18:11:42 minimezz kernel: [ 4852.340334] rtl8821ae 0000:02:00.0:
swiotlb buffer is full (sz: 9100 bytes)
Mar 16 18:11:42 minimezz kernel: [ 4852.340337] DMA: Out of SW-IOMMU space
for 9100 bytes at device 0000:02:00.0
Mar 16 18:11:42 minimezz kernel: [ 4852.358568] DMA: Out of SW-IOMMU space
for 9100 bytes at device 0000:02:00.0

I'll dig into it a bit.


-Kevin

All opinions herein are just opinions, whereas the facts are wholly
unproven.

**Comment by kevinmullican** [comment]
It seems that Ubuntu 14.04 does not include the firmware bin and the
installer in rtlwifi_new does not install it. Should I install it?


-Kevin

All opinions herein are just opinions, whereas the facts are wholly
unproven.

**Comment by kevinmullican** [comment]
Please ignore the last message, I was looking at the wrong machine. Back to
the drawing board.


-Kevin

All opinions herein are just opinions, whereas the facts are wholly
unproven.

**Comment by lwfinger** [comment]
Please supply some details on what was happening with that machine. In particular, were you using the wireless network really hard?

I have looked at the code, and I see no place where it is leaking IOMMU allocations. I have turned on IOMMU debugging on my kernel along with the IOMMU leak detection. Perhaps that will show something.

**Comment by kevinmullican** [comment]
Perhaps it is in my specific usage. I am using this machine primarily as a
wireless access point using hostapd. The machine is stable if I am not
using hostapd. The problem shows up after the following events:

1. start hostapd
2. connect to this hostapd machine from another as a wireless client
3. disconnect the client
4. wait 60-120 minutes

At this point, the machine will crash. See below for the contents of
'/etc/hostapd/hostapd.conf'

Though I am a seasoned software developer, linux hacker, and previous head
of IT, I am very ignorant in device driver land. How can I enable IOMMU
debugging (for that matter, what kernel driver tools do you like to use)?

Your implementation seems otherwise very robust and I am grateful that it
even exists. The rtl8821ae driver that ships with Ubuntu 14.04 does not
function (it loads as a module but wireless is not functional).

Thanks for you continued help,
Kevin

/etc/hostapd/hostapd.conf:

---CUT HERE---
# wifi if
interface=wlan0

# bridge
#bridge=br0

# driver
driver=nl80211

# control interface
ctrl_interface=/var/run/hostapd

# access control on the interface
ctrl_interface_group=0

# country code
country_code=US

# ssid
ssid=minimezz

# 802.11 mode
hw_mode=g

# channel
channel=3

# wpa mode
wpa=2

# wpa passphrase
wpa_passphrase=[omitted]

# Key management algorithms
wpa_key_mgmt=WPA-PSK

# Set cipher suites (encryption algorithms)
# TKIP = Temporal Key Integrity Protocol
# CCMP = AES in Counter mode with CBC-MAC
wpa_pairwise=TKIP
rsn_pairwise=CCMP

# Shared Key Authentication
auth_algs=1

# Accept all MAC address
macaddr_acl=0

# Do not hide the ssid
ignore_broadcast_ssid=0

# 802.11n
ieee80211n=1

# enable dual channel 802.11n
ht_capab=[HT40+][SMPS-STATIC][SHORT-GI-20][SHORT-GI-40][MAX-AMSDU-7935][DSSS_CCK-40]

# wmm, needed for 802.11n
wmm_enabled=1

# logging
#logger_syslog=-1
#logger_syslog_level=0
---CUT HERE---



-Kevin

All opinions herein are just opinions, whereas the facts are wholly
unproven.

**Comment by lwfinger** [comment]
Do you run NetworkManager? If so, do you see the same problem when creating an AP using the applet? I use openSUSE with a KDE desktop, and that step is very easy.

There are a couple of kernel configuration options that enable IOMMU debugging and turn on IOMMU leak tracing. To enable those, you would need to build your own kernel as no distro kernel will have enabled these - they hurt performance.

When no client is connected to an AP, the only thing that is happening is beacon transmission. Those should not leak, but I will try your hostapd configuration. Perhaps the IOMMU leak debugging will pay off there.

**Comment by kevinmullican** [comment]
Hi,

I do not use NetworkManager, because it conflicts with another application
running on the machine. I was not aware that you could setup an AP with an
applet, thanks for that. Unfortunately, that won't work in my scenario.

I appreciate your taking a look. The problem only seems to arise after
connecting and then disconnecting a client to the ap. I am capable of
building a custom kernel, if need be.

Kevin


-Kevin

All opinions herein are just opinions, whereas the facts are wholly
unproven.

**Comment by lwfinger** [comment]
I have found some problem. The connection to the hostapd AP did not finish as I do not have a bridge device configured, but the client authenticated. I had to go out and left the machine running with the log console showing. Unfortunately, the screen had blanked and the machine was dead when I returned. Nothing was logged.

I will see if I can setup some kind of debugging.

**Comment by kevinmullican** [comment]
Nice to know it's not just me. I turned off bridging in my debugging. It's
easy enough to enable bridging, just uncomment the "bridge=br0" in
hostapd.conf and setup your network accordingly (build a br0 bridge and
attach wlan0 and eth0, or appropriate, devices). You could also just NAT
via iptables too.

Did you see any errors in /var/log/syslog?


-Kevin

All opinions herein are just opinions, whereas the facts are wholly
unproven.

**Comment by lwfinger** [comment]
Nol. Actually, my system logs are viewable using the journalctl command, but any errors were lost.

Please try the patch listed below. Unfortunately, GitHub only allows attaching pictures, not text.

Index: rtlwifi_new/rtl8821ae/trx.c
===================================================================
--- rtlwifi_new.orig/rtl8821ae/trx.c
+++ rtlwifi_new/rtl8821ae/trx.c
@@ -1277,10 +1277,13 @@ bool rtl8821ae_is_tx_desc_closed(struct
         *descriptor defautly,and the own may not
         *be cleared by the hardware
         */
-       if (own)
+       if (own) {
+               if (hw_queue == BEACON_QUEUE)
+                       return true;
                return false;
-       else
+       } else {
                return true;
+       }
 }
 
 void rtl8821ae_tx_polling(struct ieee80211_hw *hw, u8 hw_queue)

**Comment by kevinmullican** [comment]
Hi,

Thanks for the this. It seems that the "+" and "-" on the diff got
converted to bullets (and perhaps a newline got munched?) on the html you
pasted in. Sadly, I'm not sure what changed. Can you send the patch as a
text file, or paste in <pre>...</pre>?

If you commit it to github, I can look there too.

Kevin


-Kevin

All opinions herein are just opinions, whereas the facts are wholly
unproven.

**Comment by lwfinger** [comment]
I just pushed the equivalent patch. It is not the complete solution, but I would like to know if the system lasts longer with this one applied.

Larry

**Comment by kevinmullican** [comment]
Did you push to master?


-Kevin

All opinions herein are just opinions, whereas the facts are wholly
unproven.

**Comment by kevinmullican** [comment]
I think I worked it out. I believe the change you wanted was:

starting at line 1277 in rtl8821ae/trx.c:

   1277          *descriptor defautly,and the own may not
   1278          *be cleared by the hardware
   1279          */
   1280         if (own) {
   1281                 if (hw_queue == BEACON_QUEUE)
   1282                         return true;
   1283                 return false;
   1284         } else {
   1285                 return true;
   1286         }


-Kevin

All opinions herein are just opinions, whereas the facts are wholly
unproven.

**Comment by alimg** [comment]
Last patch did not solve the problem for me. I am getting DMA errors after ~70 minutes of running in AP mode. My wifi is rtl8723ae, kernel: Linux arch 3.19.2-1-ARCH #1 SMP PREEMPT Wed Mar 18 16:21:02 CET 2015 x86_64 GNU/Linux

I have tried both current master branch and manually applying your patch to rtl8723ae/trx.c

**Comment by kevinmullican** [comment]
Same for me, ran for about 70 minutes in AP mode, then DMA errors
thereafter. Relevant bits pasted in:

Mar 19 13:17:17 minimezz kernel: [ 5039.723082] rtl8821ae 0000:02:00.0:
swiotlb buffer is full (sz: 9100 bytes)
Mar 19 13:17:17 minimezz kernel: [ 5039.723085] DMA: Out of SW-IOMMU space
for 9100 bytes at device 0000:02:00.0
Mar 19 13:17:17 minimezz kernel: [ 5039.728040] rtl8821ae 0000:02:00.0:
swiotlb buffer is full (sz: 9100 bytes)
Mar 19 13:17:17 minimezz kernel: [ 5039.728042] DMA: Out of SW-IOMMU space
for 9100 bytes at device 0000:02:00.0
Mar 19 13:17:17 minimezz kernel: [ 5039.926706] rtl8821ae 0000:02:00.0:
swiotlb buffer is full (sz: 9100 bytes)
Mar 19 13:17:17 minimezz kernel: [ 5039.926708] DMA: Out of SW-IOMMU space
for 9100 bytes at device 0000:02:00.0
Mar 19 13:17:17 minimezz kernel: [ 5039.928037] rtl8821ae 0000:02:00.0:
swiotlb buffer is full (sz: 9100 bytes)
Mar 19 13:17:17 minimezz kernel: [ 5039.928039] DMA: Out of SW-IOMMU space
for 9100 bytes at device 0000:02:00.0
Mar 19 13:17:17 minimezz kernel: [ 5039.930395] rtl8821ae 0000:02:00.0:
swiotlb buffer is full (sz: 9100 bytes)
Mar 19 13:17:17 minimezz kernel: [ 5039.930397] DMA: Out of SW-IOMMU space
for 9100 bytes at device 0000:02:00.0
Mar 19 13:17:17 minimezz kernel: [ 5039.933082] rtl8821ae 0000:02:00.0:
swiotlb buffer is full (sz: 9100 bytes)
Mar 19 13:17:17 minimezz kernel: [ 5039.933084] DMA: Out of SW-IOMMU space
for 9100 bytes at device 0000:02:00.0
Mar 19 13:17:17 minimezz kernel: [ 5039.956734] rtl8821ae 0000:02:00.0:
swiotlb buffer is full (sz: 9100 bytes)
Mar 19 13:17:17 minimezz kernel: [ 5039.956736] DMA: Out of SW-IOMMU space
for 9100 bytes at device 0000:02:00.0
Mar 19 13:17:17 minimezz kernel: [ 5040.132463] rtl8821ae 0000:02:00.0:
swiotlb buffer is full (sz: 9100 bytes)
Mar 19 13:17:17 minimezz kernel: [ 5040.132470] DMA: Out of SW-IOMMU space
for 9100 bytes at device 0000:02:00.0
Mar 19 13:17:17 minimezz kernel: [ 5040.133597] rtl8821ae 0000:02:00.0:
swiotlb buffer is full (sz: 9100 bytes)
Mar 19 13:17:17 minimezz kernel: [ 5040.133599] DMA: Out of SW-IOMMU space
for 9100 bytes at device 0000:02:00.0
Mar 19 13:17:17 minimezz kernel: [ 5040.135938] rtl8821ae 0000:02:00.0:
swiotlb buffer is full (sz: 9100 bytes)
Mar 19 13:17:17 minimezz kernel: [ 5040.135940] DMA: Out of SW-IOMMU space
for 9100 bytes at device 0000:02:00.0
Mar 19 13:17:17 minimezz kernel: [ 5040.138541] DMA: Out of SW-IOMMU space
for 9100 bytes at device 0000:02:00.0
Mar 19 13:17:19 minimezz kernel: [ 5041.166141] DMA: Out of SW-IOMMU space
for 9100 bytes at device 0000:02:00.0
Mar 19 13:17:19 minimezz kernel: [ 5041.168923] DMA: Out of SW-IOMMU space
for 9100 bytes at device 0000:02:00.0
Mar 19 13:17:19 minimezz kernel: [ 5041.599010] DMA: Out of SW-IOMMU space
for 9100 bytes at device 0000:02:00.0
Mar 19 13:17:19 minimezz kernel: [ 5041.771382] DMA: Out of SW-IOMMU space
for 9100 bytes at device 0000:02:00.0
Mar 19 13:17:19 minimezz kernel: [ 5041.775072] DMA: Out of SW-IOMMU space
for 9100 bytes at device 0000:02:00.0
Mar 19 13:17:19 minimezz kernel: [ 5041.962370] DMA: Out of SW-IOMMU space
for 9100 bytes at device 0000:02:00.0
Mar 19 13:17:19 minimezz kernel: [ 5041.976333] DMA: Out of SW-IOMMU space
for 9100 bytes at device 0000:02:00.0
Mar 19 13:17:19 minimezz kernel: [ 5041.994146] DMA: Out of SW-IOMMU space
for 9100 bytes at device 0000:02:00.0
Mar 19 13:17:20 minimezz kernel: [ 5042.182874] DMA: Out of SW-IOMMU space
for 9100 bytes at device 0000:02:00.0
Mar 19 13:17:20 minimezz kernel: [ 5042.185265] DMA: Out of SW-IOMMU space
for 9100 bytes at device 0000:02:00.0
Mar 19 13:17:20 minimezz kernel: [ 5042.388209] DMA: Out of SW-IOMMU space
for 9100 bytes at device 0000:02:00.0
Mar 19 13:17:20 minimezz kernel: [ 5042.393512] DMA: Out of SW-IOMMU space
for 9100 bytes at device 0000:02:00.0
Mar 19 13:17:20 minimezz kernel: [ 5042.592557] DMA: Out of SW-IOMMU space
for 9100 bytes at device 0000:02:00.0
Mar 19 13:17:20 minimezz kernel: [ 5042.597518] DMA: Out of SW-IOMMU space
for 9100 bytes at device 0000:02:00.0
Mar 19 13:17:20 minimezz kernel: [ 5042.796186] DMA: Out of SW-IOMMU space
for 9100 bytes at device 0000:02:00.0
Mar 19 13:17:40 minimezz kernel: [ 5062.551525] swiotlb_tbl_map_single: 16
callbacks suppressed
Mar 19 13:17:40 minimezz kernel: [ 5062.551529] rtl8821ae 0000:02:00.0:
swiotlb buffer is full (sz: 9100 bytes)
Mar 19 13:17:40 minimezz kernel: [ 5062.551530] DMA: Out of SW-IOMMU space
for 9100 bytes at device 0000:02:00.0

then:

Mar 19 13:20:10 minimezz kernel: [ 5212.395178] BUG: soft lockup - CPU#3
stuck for 22s! [swapper/3:0]
Mar 19 13:20:10 minimezz kernel: [ 5212.395180] Modules linked in: ctr ccm
8021q garp stp mrp llc snd_hda_codec_hdmi rfcomm bnep nls_iso8859_1
ip6table_filter ip6_tables hid_generic iptable_filter ip_tables x_tables
arc4 x86_pkg_temp_thermal intel_powerclamp coretemp kvm_intel kvm
crct10dif_pclmul crc32_pclmul ghash_clmulni_intel aesni_intel aes_x86_64
lrw gf128mul glue_helper ablk_helper snd_soc_rt5640 cryptd
snd_hda_codec_realtek snd_soc_rl6231 snd_soc_core snd_hda_codec_generic
rtl8821ae(OE) btcoexist(OE) rtl_pci(OE) rtlwifi(OE) snd_compress mac80211
btusb snd_pcm_dmaengine bluetooth usbhid snd_seq_midi snd_seq_midi_event
snd_rawmidi 6lowpan_iphc i915 cfg80211 snd_hda_intel snd_hda_controller
snd_hda_codec drm_kms_helper snd_hwdep mei_me drm lpc_ich mei snd_pcm
shpchp i2c_algo_bit snd_seq snd_seq_device snd_timer i2c_hid hid snd
soundcore video dw_dmac snd_soc_sst_acpi i2c_designware_platform parport_pc
dw_dmac_core i2c_designware_core 8250_dw ppdev spi_pxa2xx_platform lp
mac_hid parport ahci r8169 libahci mii sdhci_acpi sdhci
Mar 19 13:20:43 minimezz kernel: [ 5246.186582] NMI backtrace for cpu 3
Mar 19 13:20:43 minimezz kernel: [ 5246.186584] CPU: 3 PID: 0 Comm:
swapper/3 Tainted: G           OE 3.16.0-31-generic #43~14.04.1-Ubuntu
Mar 19 13:20:43 minimezz kernel: [ 5246.186585] Hardware name: GIGABYTE
M4HM87P-00/M4HM87P-00, BIOS F5 06/23/2014
Mar 19 13:20:43 minimezz kernel: [ 5246.186586] task: ffff8802132e28c0 ti:
ffff880212d18000 task.ti: ffff880212d18000
Mar 19 13:20:43 minimezz kernel: [ 5246.186587] RIP:
0010:[<ffffffffc084804f>]  [<ffffffffc084804f>] pci_read32_sync+0xf/0x20
[rtl_pci]
Mar 19 13:20:43 minimezz kernel: [ 5246.186593] RSP: 0018:ffff88021eac3d20
 EFLAGS: 00000086
Mar 19 13:20:43 minimezz kernel: [ 5246.186594] RAX: 0000000000000000 RBX:
ffff880210ab1840 RCX: 00000000000050ee
Mar 19 13:20:43 minimezz kernel: [ 5246.186595] RDX: 0000000000000000 RSI:
ffffc90004ed00b8 RDI: ffff880210ab1840
Mar 19 13:20:43 minimezz kernel: [ 5246.186595] RBP: ffff88021eac3d20 R08:
0000000000000002 R09: ffff880215c00000
Mar 19 13:20:43 minimezz kernel: [ 5246.186596] R10: 0000000000000000 R11:
0000000000000003 R12: ffff880210ab0680
Mar 19 13:20:43 minimezz kernel: [ 5246.186597] R13: ffff880210ab18dc R14:
0000000000000002 R15: ffff880212d1bdc8
Mar 19 13:20:43 minimezz kernel: [ 5246.186597] FS:  0000000000000000(0000)
GS:ffff88021eac0000(0000) knlGS:0000000000000000
Mar 19 13:20:43 minimezz kernel: [ 5246.186598] CS:  0010 DS: 0000 ES: 0000
CR0: 0000000080050033
Mar 19 13:20:43 minimezz kernel: [ 5246.186599] CR2: 00007fedbb083df8 CR3:
0000000002c13000 CR4: 00000000001407e0
Mar 19 13:20:43 minimezz kernel: [ 5246.186599] Stack:
Mar 19 13:20:43 minimezz kernel: [ 5246.186600]  ffff88021eac3d38
ffffffffc088fb94 ffff880210ab1840 ffff88021eac3d80
Mar 19 13:20:43 minimezz kernel: [ 5246.186601]  ffffffffc084a713
00000f3e899c7eca 0000000000000000 ffff880212396c00
Mar 19 13:20:43 minimezz kernel: [ 5246.186602]  0000000000000033
ffff880212396c00 0000000000000000 ffff880212d1bdc8
Mar 19 13:20:43 minimezz kernel: [ 5246.186604] Call Trace:
Mar 19 13:20:43 minimezz kernel: [ 5246.186604]  <IRQ>
Mar 19 13:20:43 minimezz kernel: [ 5246.186605]  [<ffffffffc088fb94>]
rtl8821ae_disable_interrupt+0x64/0x70 [rtl8821ae]
Mar 19 13:20:43 minimezz kernel: [ 5246.186617]  [<ffffffffc084a713>]
_rtl_pci_interrupt+0x73/0x8e0 [rtl_pci]
Mar 19 13:20:43 minimezz kernel: [ 5246.186624]  [<ffffffff820cb2ee>]
handle_irq_event_percpu+0x3e/0x1a0
Mar 19 13:20:43 minimezz kernel: [ 5246.186631]  [<ffffffff820cb48d>]
handle_irq_event+0x3d/0x60
Mar 19 13:20:43 minimezz kernel: [ 5246.186636]  [<ffffffff820ce047>]
handle_edge_irq+0x77/0x130
Mar 19 13:20:43 minimezz kernel: [ 5246.186640]  [<ffffffff820155ee>]
handle_irq+0x1e/0x30
Mar 19 13:20:43 minimezz kernel: [ 5246.186641]  [<ffffffff8276e46f>]
do_IRQ+0x4f/0xf0
Mar 19 13:20:43 minimezz kernel: [ 5246.186643]  [<ffffffff8276c2ed>]
common_interrupt+0x6d/0x6d
Mar 19 13:20:43 minimezz kernel: [ 5246.186645]  [<ffffffff8207302b>] ?
__do_softirq+0x8b/0x2e0
Mar 19 13:20:43 minimezz kernel: [ 5246.186646]  [<ffffffff82072fc0>] ?
__do_softirq+0x20/0x2e0
Mar 19 13:20:43 minimezz kernel: [ 5246.186647]  [<ffffffff82073555>]
irq_exit+0x105/0x110
Mar 19 13:20:43 minimezz kernel: [ 5246.186648]  [<ffffffff8276e478>]
do_IRQ+0x58/0xf0
Mar 19 13:20:43 minimezz kernel: [ 5246.186650]  [<ffffffff8276c2ed>]
common_interrupt+0x6d/0x6d
Mar 19 13:20:43 minimezz kernel: [ 5246.186650]  <EOI>
Mar 19 13:20:43 minimezz kernel: [ 5246.186651]  [<ffffffff825fc032>] ?
cpuidle_enter_state+0x52/0xc0
Mar 19 13:20:43 minimezz kernel: [ 5246.186656]  [<ffffffff825fc028>] ?
cpuidle_enter_state+0x48/0xc0
Mar 19 13:20:43 minimezz kernel: [ 5246.186663]  [<ffffffff825fc157>]
cpuidle_enter+0x17/0x20
Mar 19 13:20:43 minimezz kernel: [ 5246.186665]  [<ffffffff820b52cd>]
cpu_startup_entry+0x31d/0x450
Mar 19 13:20:43 minimezz kernel: [ 5246.186667]  [<ffffffff8204534d>]
start_secondary+0x21d/0x2e0
Mar 19 13:20:43 minimezz kernel: [ 5246.186667] Code: 03 b7 d8 09 00 00 48
89 e5 66 8b 06 5d c3 66 66 66 66 66 2e 0f 1f 84 00 00 00 00 00 55 89 f6 48
03 b7 d8 09 00 00 48 89 e5 8b 06 <5d> c3 66 66 66 66 66 66 2e 0f 1f 84 00
00 00 00 00 55 89 f6 48


Hope this helps,
Kevin




-Kevin

All opinions herein are just opinions, whereas the facts are wholly
unproven.

**Comment by lwfinger** [comment]
Thanks. Yes, it helps, but likely more helpful was some information I got overnight from the Realtek engineer. It seems that completion of a beacon transmission does not generate an interrupt, thus the TX interrupt service routine will not be called. As a result, my fixup attempt is useless. The cleanup after the beacon is sent is done in a tasklet completion routine, whichdid not unmap the DMA buffer. I just committed a change to master that should fix the problem. I have not had a chance to test it yet so we will all test in parallel.

**Comment by kevinmullican** [comment]
Nice. I'll give it a go.

Kevin

**Comment by lwfinger** [comment]
It has worked for about 3 hours here. I added some additional printouts here, before this patch, it was leaking one DMA mapping for every beacon. At 0.1 sec between beacons, that leaked a lot in an hour.

If you are willing to have your E-mail address published in an open list, I will list you as Reported-by and Tested-by as appropriate for the kernel patch.

**Comment by alimg** [comment]
Got 7 hours without any problem and counting on. Thank you for solving this!

**Comment by kevinmullican** [comment]
Working for me so far. You can put my email out there.

Kevin

**Comment by lwfinger** [comment]
I need your E-mail address. Send it to Larry.Finger@lwfinger.net

**Comment by lambertdw** [comment]
Hi,

My Thinkpad W540 Ubuntu 14.10 system works considerably better now.
Next I'll try to enable bluetooth.


lshw
            *-network
                 description: Wireless interface
                 product: RTL8192EE PCIe Wireless Network Adapter
                 vendor: Realtek Semiconductor Co., Ltd.
                 physical id: 0
                 bus info: pci@0000:03:00.0
                 logical name: wlan0
                 version: 00
                 serial: 74:29:af:9a:21:cb
                 width: 64 bits
                 clock: 33MHz
                 capabilities: pm msi pciexpress bus_master cap_list 
ethernet physical wireless
                 configuration: broadcast=yes driver=rtl8192ee 
driverversion=3.16.0-31-generic firmware=N/A ip=192.168.0.195 latency=0 
link=yes multicast=yes wireless=IEEE 802.11bgn
                 resources: irq:46 ioport:4000(size=256) 
memory:b2800000-b2803fff



iwconfig

wlan0     IEEE 802.11bgn  ESSID:"nice"
           Mode:Managed  Frequency:2.462 GHz  Access Point: 
00:21:91:03:9E:55
           Bit Rate=144.4 Mb/s   Tx-Power=20 dBm
           Retry short limit:7   RTS thr=2347 B   Fragment thr:off
           Power Management:off
           Link Quality=70/70  Signal level=-32 dBm
           Rx invalid nwid:0  Rx invalid crypt:0  Rx invalid frag:0
           Tx excessive retries:0  Invalid misc:537   Missed beacon:0


ifconfig

wlan0     Link encap:Ethernet  HWaddr 74:29:af:9a:21:cb
           inet addr:192.168.0.195  Bcast:192.168.0.255 Mask:255.255.255.0
           inet6 addr: fe80::7629:afff:fe9a:21cb/64 Scope:Link
           UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
           RX packets:253698 errors:0 dropped:0 overruns:0 frame:0
           TX packets:178645 errors:0 dropped:0 overruns:0 carrier:0
           collisions:0 txqueuelen:1000
           RX bytes:312942865 (312.9 MB)  TX bytes:31081177 (31.0 MB)



Thanks, Dave Lambert



On 03/21/2015 12:34 PM, lwfinger wrote:My

**Comment by lwfinger** [comment]
What does "lsusb" show for the USB ID for the BT part of your RTL8192EE. I was not aware that the 8192EE had a BT part.

**Comment by lwfinger** [comment]
Device 0bda:8761 is handled by the driver in the kernel branch of http://github.com/lwfinger/rtl8723au_bt.git. FYI, my RTL8192EE does not have a BT device associated with it.

**Comment by lwfinger** [comment]
A completely new version of the drivers has just been pushed. It fixes a lot of things. As a result, I am closing all open issues. If you get a new one, please create a new issue.

Note: The new driver fixes the problems with power save. It is recommended that you clear all ips, swlps, and fwlps options.
